File: src/main/java/io/vproxy/vfx/ui/shapes/MovablePoint.java
package io.vproxy.vfx.ui.shapes;

import io.vproxy.vfx.control.drag.DragHandler;
import io.vproxy.vfx.entity.Point;
import io.vproxy.vfx.manager.font.FontManager;
import io.vproxy.vfx.manager.font.FontUsages;
import io.vproxy.vfx.util.FXUtils;
import javafx.scene.Cursor;
import javafx.scene.Group;
import javafx.scene.control.Label;
import javafx.scene.paint.Color;
import javafx.scene.shape.Circle;

public class MovablePoint extends Group {
    public MovablePoint(String labelText) {
        var point = new Circle(5);
        point.setStrokeWidth(2);
        point.setStroke(Color.RED);
        point.setFill(Color.TRANSPARENT);
        var dot = new Circle(2);
        dot.setFill(Color.RED);
        dot.setStrokeWidth(0);
        var label = new Label(labelText) {{
            FontManager.get().setFont(FontUsages.movableShapeLabel, this);
        }};
        point.setCursor(Cursor.MOVE);
        label.setTextFill(Color.RED);
        var wh = FXUtils.calculateTextBounds(label);
        label.setLayoutX(-wh.getWidth() / 2);
        label.setLayoutY(10);

        var handler = new DragHandler() {
            @Override
            protected void set(double x, double y) {
                setLayoutX(x);
                setLayoutY(y);
            }

            @Override
            protected double[] get() {
                return new double[]{getLayoutX(), getLayoutY()};
            }
        };

        var pointAndDotGroup = new Group(point, dot);
        pointAndDotGroup.setOnMousePressed(handler);
        pointAndDotGroup.setOnMouseDragged(handler);

        getChildren().addAll(label, pointAndDotGroup);
    }

    public Point makePoint() {
        var point = new Point();
        point.x = getLayoutX();
        point.y = getLayoutY();
        return point;
    }

    public void from(Point point) {
        setX(point.x);
        setY(point.y);
    }

    public void setX(double x) {
        setLayoutX(x);
    }

    public void setY(double y) {
        setLayoutY(y);
    }
}

File: src-ts/src/ui/shapes/MovablePoint.ts
import { Group } from '../../javafx/Pane.js';
import { Circle } from '../../javafx/shapes.js';
import { Label } from '../../javafx/Label.js';
import { Color } from '../../javafx/color.js';
import { Point } from '../../entity/Point.js';
import { FontManager } from '../../manager/font/FontManager.js';
import { FontUsages } from '../../manager/font/FontUsages.js';
import { FXUtils } from '../../util/FXUtils.js';
import { DragHandler } from '../../control/drag/DragHandler.js';

export class MovablePoint extends Group {
  constructor(labelText: string) {
    super();
    const point = new Circle(5);
    point.setStrokeWidth(2);
    point.setStroke(Color.RED);
    point.setFill(Color.TRANSPARENT);
    const dot = new Circle(2);
    dot.setFill(Color.RED);
    dot.setStrokeWidth(0);
    const label = new Label(labelText);
    FontManager.get().setFont(FontUsages.movableShapeLabel, label);
    point.setCursor('move');
    label.setTextFill(Color.RED);
    const wh = FXUtils.calculateTextBounds(label);
    label.setLayoutX(-wh.width / 2);
    label.setLayoutY(10);

    const self = this;
    const handler = new (class extends DragHandler {
      protected set(x: number, y: number): void {
        self.setLayoutX(x);
        self.setLayoutY(y);
      }

      protected get(): [number, number] {
        return [self.getLayoutX(), self.getLayoutY()];
      }
    })();

    const pointAndDotGroup = new Group();
    pointAndDotGroup.getChildren().addAll(point, dot);
    pointAndDotGroup.setOnMousePressed(handler.handlePressed);
    pointAndDotGroup.setOnMouseDragged(handler.handleDragged);

    this.getChildren().addAll(label, pointAndDotGroup);
  }

  makePoint(): Point {
    const point = new Point();
    point.x = this.getLayoutX();
    point.y = this.getLayoutY();
    return point;
  }

  from(point: Point): void {
    this.setX(point.x);
    this.setY(point.y);
  }

  setX(x: number): void {
    this.setLayoutX(x);
  }

  setY(y: number): void {
    this.setLayoutY(y);
  }
}
